core/metadata.h

```diff
@@ -3,73 +3,26 @@
-    promise(offset >= 0);                                   \
-    struct snbuf *_pkt = (pkt);                             \
-    uintptr_t addr = (uintptr_t)(_pkt->_metadata + offset); \
-    (type *)addr;                                           \
-  })
-
-#define _get_attr_with_offset(offset, pkt, type) \
-  ({ *(_ptr_attr_with_offset(offset, pkt, type)); })
-
-#define _set_attr_with_offset(offset, pkt, type, val)   \
-  ((void)({                                             \
-    type _val = (val);                                  \
-    *(_ptr_attr_with_offset(offset, pkt, type)) = _val; \
-  }))
-
-// Safe version.
-#define ptr_attr_with_offset(offset, pkt, type)             \
-  ({                                                        \
-    bess::metadata::mt_offset_t _offset = (offset);         \
-    bess::metadata::IsValidOffset(_offset)                  \
-        ? (type *)_ptr_attr_with_offset(_offset, pkt, type) \
-        : (type *)nullptr;                                  \
-  })
-
-#define get_attr_with_offset(offset, pkt, type)                                              \
-  ({                                                                                         \
-    static type _zeroed;                                                                     \
-    bess::metadata::mt_offset_t _offset = (offset);                                          \
-    bess::metadata::IsValidOffset(_offset) ? (type)_get_attr_with_offset(_offset, pkt, type) \
-                             : (type)_zeroed;                                                \
-  })
-
-#define set_attr_with_offset(offset, pkt, type, val)  \
-  ((void)({                                           \
-    bess::metadata::mt_offset_t _offset = (offset);   \
-    if (bess::metadata::IsValidOffset(_offset))        \
-      _set_attr_with_offset(_offset, pkt, type, val); \
-  }))
-
-// Slowest but easiest.
-#define ptr_attr(module, attr_id, pkt, type) \
-  ptr_attr_with_offset(module->attr_offsets[attr_id], pkt, type)
-
-#define get_attr(module, attr_id, pkt, type) \
-  get_attr_with_offset(module->attr_offsets[attr_id], pkt, type)
-
-#define set_attr(module, attr_id, pkt, type, val) \
-  set_attr_with_offset(module->attr_offsets[attr_id], pkt, type, val)
+#include <gtest/gtest_prod.h>
+
+class Module;
 
 namespace bess {
 namespace metadata {
 
-static const int kMetadataAttrMaxSize = 32; // In bytes, per attribute.
+static const int kMetadataAttrMaxSize = 32;  // In bytes, per attribute.
 
-static const int kMaxAttrsPerModule = 16; // Max number of attributes per module.
+static const int kMaxAttrsPerModule =
+    16;  // Max number of attributes per module.
 
-static const int kMetadataTotalSize = 96; // Total size of all metadata in bytes.
+static const int kMetadataTotalSize =
+    96;  // Total size of all metadata in bytes.
 
 // Normal offset values are 0 or a positive value.
 typedef int8_t mt_offset_t;
```

Comment:
What is the non-normal offset value here, and when is it used? (can we use unsigned?)